lib/multibuild/build.ts

```diff
@@ -0,0 +1,346 @@
+import { getManifest, MEDIATYPE_MANIFEST_LIST_V2 } from './manifests';
+
+function taskHooks(
+	task: BuildTask,
+	docker: Dockerode,
+	resolve: (image: LocalImage) => void,
+): BuildHooks {
+	let startTime: number;
+
+	const setImageProperties = (
+		image: LocalImage,
+		layers: string[],
+		fromTags: FromTagInfo[],
+	) => {
+		image.layers = layers;
+		image.baseImageTags = fromTags;
+		image.startTime = startTime;
+		image.endTime = Date.now();
+		image.dockerfile = task.dockerfile;
+		image.projectType = task.projectType;
+	};
+
+	return {
+		buildSuccess: (
+			imageId: string,
+			layers: string[],
+			fromTags: FromTagInfo[],
+		) => {
+			const tag = task.tag != null ? task.tag : imageId;
+			const image = new LocalImage(docker, tag, task.serviceName, {
+				external: false,
+				successful: true,
+			});
+			setImageProperties(image, layers, fromTags);
+			resolve(image);
+		},
+		buildFailure: (error: Error, layers: string[], fromTags: FromTagInfo[]) => {
+			const image = new LocalImage(
+				docker,
+				layers[layers.length - 1],
+				task.serviceName,
+				{ external: false, successful: false },
+			);
+			setImageProperties(image, layers, fromTags);
+			image.error = error;
+			resolve(image);
+		},
+		buildStream: (stream: Stream.Duplex) => {
+			startTime = Date.now();
+			if (_.isFunction(task.streamHook)) {
+				task.streamHook(stream);
+			}
+
+			task.buildStream!.pipe(stream);
+		},
+	};
+}
+
+const generateBuildArgs = (
+	task: BuildTask,
+	userArgs?: Dictionary<string>,
+): { buildargs?: Dictionary<string> } => {
+	return {
+		buildargs: { ...task.args, ...userArgs },
+	};
+};
+
+const generateLabels = (task: BuildTask): { labels?: Dictionary<string> } => {
+	return {
+		labels: task.labels,
+	};
+};
+
+/**
+ * Given a build task which is primed with the necessary input, perform either
+ * a build or a docker pull, and return this as a LocalImage.
+ *
+ * @param task The build task to perform
+ * @param docker The handle to the docker daemon
+ * @return a promise which resolves to a LocalImage which points to the produced image
+ */
+export async function runBuildTask(
+	task: BuildTask,
+	docker: Dockerode,
+	registrySecrets: RegistrySecrets,
+	secrets?: SecretsPopulationMap,
+	buildArgs?: Dictionary<string>,
+): Promise<LocalImage> {
+	// Determine how we should handle the `platform` flag.
+	// This will be a combination of factors:
+	//
+	// * _ Does the Docker version support it? _
+	//   NOTE: docker api version 1.34 actually introduced platform to the
+	//         api but it was broken until fixed in 1.38
+	//         https://github.com/moby/moby/commit/7f334d3acfd7bfde900e16e393662587b9ff74a1
+	//
+	// * _ Do the images support it? _
+	//   In order to support `platform`, all images used in the Dockerfile must use the same
+	//   architecture.  Determining this is problematic.  See comments in getRecommendedPlatformHandling
+
+	const usePlatformOption: boolean =
```

Comment:
FYI:  Once this merges, in order to unblock the (formerly) multibuild issue, I'll PR a change to allow the default `--platform` behavior here with a new option on the build task `task.useDefaultPlatformForMultiarchBaseImages` and then 

```
	const usePlatformOption: boolean =
		!!task.dockerPlatform &&
		semver.satisfies(
			semver.coerce((await docker.version()).ApiVersion) || '0.0.0',
			'>=1.38.0',
		) &&
		(task.useDefaultPlatformForMultiarchBaseImages ||
			(await checkAllowDockerPlatformHandling(task, docker)));
```